src/mumble/JackAudio.cpp

```diff
@@ -1059,6 +1125,84 @@ bool JackAudioOutput::process(const jack_nframes_t frames) {
 	return true;
 }
 
+void JackAudioOutput::prepareOutputBuffers(unsigned int frameCount, QList< AudioOutputBuffer * > &qlMix,
+										   QList< AudioOutputBuffer * > &qlDel) {
+	ServerHandlerPtr sh = Global::get().sh;
+	VoiceRecorderPtr recorder;
+	if (sh) {
+		recorder = Global::get().sh->recorder;
+	}
+
+	// Register user ports based on who is currently present in qmOutputs and route their audio to JACK.
+	// Don't care if users get removed for now. TODO: maybe at some point we do.
+	QMultiHash< const ClientUser *, AudioOutputBuffer * >::const_iterator it = qmOutputs.constBegin();
+	while (it != qmOutputs.constEnd()) {
+		const ClientUser *user   = it.key();
+		AudioOutputBuffer *audio = it.value();
+
+		if (!user || !recorder || (recorder && !recorder->isTransportEnabled())) {
+			if (audio->prepareSampleBuffer(frameCount)) {
+				qlMix.append(audio);
+			} else {
+				qlDel.append(audio);
+			}
+
+			++it;
+			continue;
+		}
+
+		QString qsPortName = user->qsName;
+		qsPortName         = qsPortName.prepend("user_");
+		if (!userPorts[qsPortName]) {
```

Comment:
Maybe using `operator[]` here can be dangerous as it will auto-insert an entry if the key doesn't exist yet?

Reply:
Yes, this does auto-insert a key. But, it initializes the mapped value to whatever is the default for the type (i.e zero-valued/null in the case of a pointer).

Therefore no, this is not dangerous because this is standardized compiler behavior (we will never read a garbage value from an auto-added key)

Reply:
The "danger" I was referring to is that different parts of the code could check for the existence of keys and then do false deductions because of an auto-added key.
Another thing to consider here is whether the auto-insertion is something you actually _want_ to happen. If not, then memory-wise it would probably be a good idea to not insert entries for all user names. But if this is something that you actually make good use of, it's fine of course

Reply:
Understood, it's been a hot minute so it's possible I had a more clever reason to write it this way. I'll check into it and fix the behavior if it is problematic in the way you're describing.